sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java

```diff
@@ -150,47 +143,109 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
+        List<URI> requestLocationsList = new ArrayList<>();
+        final List<URI> orderedEffectiveEndpointsList = getApplicableEndPoints(request);
+        if (orderedEffectiveEndpointsList != null) {
+            int i = 0;
+            while (i < orderedEffectiveEndpointsList.size()) {
+                if (orderedEffectiveEndpointsList.get(i) == null) {
+                    orderedEffectiveEndpointsList.remove(i);
+                } else {
+                    i++;
+                }
             }
         }
 
-        // If the above conditions are not met, then we will just return the original request
-        if (monoList.isEmpty()) {
-            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
+        if (orderedEffectiveEndpointsList == null || orderedEffectiveEndpointsList.size() == 0) {
+            // If no endpoints are available just return original Mono
+            return getStoreResponseMono(request, forceRefreshAndTimeout);
         }
 
-        return Mono.firstWithValue(monoList);
+        orderedEffectiveEndpointsList
+            .forEach(locationURI -> {
+                RxDocumentServiceRequest newRequest = request.clone();
+                newRequest.requestContext.routeToLocation(locationURI);
+                requestLocationsList.add(locationURI);
+                if (monoList.isEmpty()) {
+                    monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout));
+                } else {
+                    monoList
+                        .add(
+                            getStoreResponseMono(newRequest, forceRefreshAndTimeout)
+                                .delaySubscription(((ThresholdBasedAvailabilityStrategy) strategy)
+                                    .getThreshold()
+                                    .plus(((ThresholdBasedAvailabilityStrategy) strategy)
+                                        .getThresholdStep()
+                                        .multipliedBy(monoList.size() - 1))
+                                ));
+                }
+            });
+
+        return Mono
+            .firstWithValue(monoList)
+            .onErrorMap(throwable -> {
+                Throwable exception = Exceptions.unwrap(throwable);
+
+                if (exception instanceof NoSuchElementException) {
+
+                    List<Throwable> innerThrowables = Exceptions
+                        .unwrapMultiple(exception.getCause());
+
+                    int index = 0;
+                    for (Throwable innerThrowable : innerThrowables) {
+                        Throwable innerException = Exceptions.unwrap(innerThrowable);
+
+                        // collect latest CosmosException instance bubbling up for a region
+                        if (innerException instanceof CosmosException) {
+                            return Utils.as(innerException, CosmosException.class);
+                        } else if (innerException instanceof NoSuchElementException) {
```

Comment:
What is the possibility of nested NoSuchElementExceptions?

Reply:
A NoSucheElementException from Mono.firstWithValue would have a CompositeException with exactly as many exception as Monos (same order)
Inner NoSuchElementException would indicate Mono received empty signal
Mono.error (Exception)
Mono.empty NOSuchElementException
Mono with value - we would not even see the NoSuchElementException in the first place